### PR TITLE
add upload for rich output file

### PR DESCRIPTION
**Pull request by talbe**

### CODE REVIEW

File: output/output.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+func UploadRichOutputFile(ctx context.Context, content io.Reader, contentType int) error {
+	var uploadContentType string
+	var validContentType bool
+	if uploadContentType, validContentType = contentTypes[contentType]; !validContentType {
+		return fmt.Errorf("upload rich output file: bad content type %d", contentType)
+	}
+
+	client := &http.Client{}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPut, os.Getenv(RichOutputURL), content)
```
Comment (talbe) [inline output/output.go:40]: Not sure about the os.Getenv - WDYT?

File: output/output.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+var contentTypes = map[int]string {
+	ContentTypeBinary: "application/octet-stream",
+	ContentTypeForm: "application/x-www-form-urlencoded",
+	ContentTypeJSON: "application/json",
+	ContentTypeHTML: "text/html; charset=utf-8",
+	ContentTypeText: "text/plain; charset=utf-8",
+	ContentTypeMarkdown: "text/markdown",
+}
+
+func UploadRichOutputFile(ctx context.Context, content io.Reader, contentType int) error {
```
Comment (kostyay) [inline output/output.go:31]:
```suggestion
func Upload(ctx context.Context, dst string, contentType ContentType, content io.Reader) error {
```

File: output/output.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	var uploadContentType string
+	var validContentType bool
+	if uploadContentType, validContentType = contentTypes[contentType]; !validContentType {
+		return fmt.Errorf("upload rich output file: bad content type %d", contentType)
```
Comment (kostyay) [inline output/output.go:35]:
```suggestion
		return fmt.Errorf("invalid content type %d", contentType)
```

File: output/output.go (hunk 1 above)
Comment (kostyay) [reply output/output.go:40]: > Not sure about the os.Getenv - WDYT?

```suggestion
	req, err := http.NewRequestWithContext(ctx, http.MethodPut, dst, content)
```

File: output/output.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	client := &http.Client{}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPut, os.Getenv(RichOutputURL), content)
+	if err != nil {
+		return fmt.Errorf("upload rich output file: %w", err)
```
Comment (kostyay) [inline output/output.go:42]:
```suggestion
		return fmt.Errorf("new http request: %w", err)
```

File: output/output.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	req.Header.Add("content-type", uploadContentType)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("upload rich output file: %w", err)
```
Comment (kostyay) [inline output/output.go:49]:
```suggestion
		return fmt.Errorf("upload: %w", err)
```

File: output/output.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+const (
+	ContentTypeBinary = 1
+	ContentTypeForm = 2
+	ContentTypeJSON = 3
+	ContentTypeHTML = 4
+	ContentTypeText = 5
+	ContentTypeMarkdown = 6
+
+	RichOutputURL = "RICH_OUTPUT_URL"
```
Comment (kostyay) [inline output/output.go:19]:
```suggestion
	envRichOutputURL = "RICH_OUTPUT_URL"
```

File: output/output.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+var contentTypes = map[int]string {
+	ContentTypeBinary: "application/octet-stream",
+	ContentTypeForm: "application/x-www-form-urlencoded",
+	ContentTypeJSON: "application/json",
+	ContentTypeHTML: "text/html; charset=utf-8",
+	ContentTypeText: "text/plain; charset=utf-8",
+	ContentTypeMarkdown: "text/markdown",
+}
+
```
Comment (kostyay) [inline output/output.go:30]:
```go
func UploadRichOutput(ctx context.Context,  t ContentType, content io.Reader) error {
    outputDst := os.Getenv(envRichOutputURL)
    if outputDst == "" {
        return error
    }
    return Upload(ctx, os.Getenv(envRichOutputURL), contentType, content)
}
```

**Review by kostyay — CHANGES_REQUESTED** [review]
see comments + add a test

File: output/output.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	"os"
+)
+
+const (
+	ContentTypeBinary = 1
+	ContentTypeForm = 2
+	ContentTypeJSON = 3
+	ContentTypeHTML = 4
+	ContentTypeText = 5
+	ContentTypeMarkdown = 6
```
Comment (alongat) [inline output/output.go:17]: Keep only the data we currently fully support

File: output/output.go [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+	req.Header.Add("content-type", uploadContentType)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("upload rich output file: %w", err)
+	}
+
+	defer resp.Body.Close()
+
```
Comment (oriser) [inline output/output.go:53]: Check the status code of the response here.
The http can return no error, but the status should be 20X (StatusCode/100 == 2)

File: upload/upload.go [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+const envRichOutputURL = "RICH_OUTPUT_URL"
+
+func RichOutput(ctx context.Context, content io.Reader, ct contentType) error {
+	uploadContentType, ok := contentTypes[ct]
+	if !ok {
+		return fmt.Errorf("invalid content type %s", ct)
+	}
+
+	richOutputURL, ok := os.LookupEnv(envRichOutputURL)
```
Comment (kostyay) [inline upload/upload.go:33]: check richOutputURL != ""

File: upload/upload.go [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+	}
+
+	req.Header.Add("content-type", uploadContentType)
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("upload rich output file: %w", err)
+	}
+	if resp.StatusCode/2 != 2 {
```
Comment (kostyay) [inline upload/upload.go:50]:
```suggestion
	if resp.StatusCode/100 != 2 {
```

**Review by kostyay — APPROVED** [review]

**Comment by kostyay** [comment]
see comment + add test

File: upload/upload.go [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+	ContentTypeMarkdown contentType = iota
+)
+
+var contentTypes = map[contentType]string{
+	ContentTypeMarkdown: "text/markdown",
+}
+
+func (c contentType) String() string {
+	return contentTypes[c]
+}
```
Comment (oriser) [inline upload/upload.go:23]: noice

File: upload/upload.go [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("upload rich output file: %w", err)
+	}
+	if resp.StatusCode/100 != 2 {
+		return fmt.Errorf("bad http status code receieved: %d", resp.StatusCode)
+	}
+
+	defer resp.Body.Close()
```
Comment (vic3lord) [inline upload/upload.go:54]:
```suggestion
	}	
	defer resp.Body.Close()
	
	if resp.StatusCode/100 != 2 {
		return fmt.Errorf("bad http status code receieved: %d", resp.StatusCode)
	}
```

Comment (vic3lord) [reply upload/upload.go:54]: Usually the `defer` comes first, right after the error of the action you tried to do 

Example: 

os.Open(file)
if err != nil {
}
defer f.Close()

**Pull request merged**